 # 4.6.2 Combining & Exporting Data
# 
This script contains the following points:# 
1. Import the data sets into Jupyte# r
2. Check the dimensions of the imported datafram# es
3. Determine a suitable way to combine the orders_products_combined dataframe with the products data # set
4. Confirm the results of the merge using the merge
# 





5. Export the newly created dataframe as ords_prods_merge in a suitable format

# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import the data sets into Jupyter

In [2]:
path = r'C:\Users\João Pedro\MAR-Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\João Pedro\\MAR-Instacart Basket Analysis'

# Import dataset orders_products_combined.pkl

In [4]:
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

# Import dataset  orders checked

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

# Check the dimensions od the imported dataframe

In [6]:
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [7]:
df_ords_prods_combined.shape

(32434489, 11)

In [8]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

# Drop unnecessary columns

In [9]:
df_prods = df_prods.drop(['Unnamed: 0'], axis=1)

# check output

In [10]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [11]:
df_prods.shape

(49672, 5)

#  Determine a suitable way to combine the orders_products_combined dataframe with the products data set

## For this project, it is recommended to merge the df_ords_prods_combined dataframe with the df_prods dataframe based on their shared "order_id" column. Despite their distinct shapes, an inner join is advisable for seamless integration. ##

In [12]:
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

ValueError: Cannot use name of an existing column for indicator column

## The issue lies in the fact that the "merge" column already exists in the df_ords_prods_combined dataframe. Consequently, I need to remove it from the dataframe before proceeding with the merging procedure. ##

# Drop unnecessary columns

In [13]:
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)

# Check output

In [14]:
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [15]:
df_ords_prods_combined.shape

(32434489, 10)

# Merge the updated dataframes

In [16]:
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [17]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  
0                     3.0        False                  5          0   both  
1                    20.0        False                  1          1   both  
2                     6.0        False                 20          0   both  
3                     NaN         True                 10          0   both  
4                     9.0        False                 11          1   both

In [18]:
df_ords_prods_merge.shape

(32404859, 15)

# CONFIRM THE RESULTS OF THE MERGE USING THE MERGE FLAG

In [20]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

## After the merge, the resultant dataframe comprises 32,404,859 rows, with each row containing information from both input data sets due to the utilization of an inner join for the project's requirements. ##

 # Export the newly created dataframe as ords_prods_merge in a suitable format 

In [21]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))